File: Common/Utilities.Reflection.vb
Imports System.Runtime.CompilerServices
Imports System.Text
Imports Media = System.Windows.Media
Imports Reflect = System.Reflection
Imports SIO = System.IO

Namespace Common

    Partial Class Utilities

        Public Class Reflection

            Private Sub New()
                Throw New NotSupportedException()
            End Sub

            Public Shared Function GetAllAccessibleAssemblies() As IEnumerable(Of Reflect.Assembly)
                Return Reflect.Assembly.GetEntryAssembly().GetRecursiveReferencedAssemblies()
            End Function

            Public Shared Iterator Function GetAllMethods(Optional ByVal Assemblies As IEnumerable(Of Reflect.Assembly) = Nothing) As IEnumerable(Of Reflect.MethodInfo)
                If Assemblies Is Nothing Then
                    Assemblies = GetAllAccessibleAssemblies()
                End If
                For Each Ass In Assemblies
                    For Each Type As Type In Ass.GetTypes()
                        For Each Method As Reflect.MethodInfo In Type.GetMethods(Reflect.BindingFlags.Static Or Reflect.BindingFlags.Instance Or Reflect.BindingFlags.Public Or Reflect.BindingFlags.NonPublic)
                            Yield Method
                        Next
                    Next
                Next
            End Function

            Public Shared Iterator Function GetAllTypes(Optional ByVal Assemblies As IEnumerable(Of Reflect.Assembly) = Nothing) As IEnumerable(Of Type)
                If Assemblies Is Nothing Then
                    Assemblies = GetAllAccessibleAssemblies()
                End If
                For Each Ass In Assemblies
                    For Each Type As Type In Ass.GetTypes()
                        Yield Type
                    Next
                Next
            End Function

            Public Shared Iterator Function GetAllTypesDerivedFrom(ByVal Base As Type, Optional ByVal Assemblies As IEnumerable(Of Reflect.Assembly) = Nothing) As IEnumerable(Of Type)
                If Assemblies Is Nothing Then
                    Assemblies = GetAllAccessibleAssemblies()
                End If
                For Each Type As Type In GetAllTypes(Assemblies)
                    If Base.IsAssignableFrom(Type) Then
                        Yield Type
                    End If
                Next
            End Function

            <Sample()>
            Public Shared Sub GetAllTypesDerivedFrom(ByVal Base As Cecil.TypeDefinition, Optional ByVal Assembly As Reflect.Assembly = Nothing)
                Dim Helper = CecilHelper.Instance

                If Assembly Is Nothing Then
                    Assembly = Reflect.Assembly.GetEntryAssembly()
                End If

                For Each A In Helper.GetReferencedAssemblies(Helper.Convert(Assembly))
                    For Each M In A.Modules
                        For Each T In M.Types
                            If Helper.IsBaseTypeOf(Base, T) Then
                                Console.WriteLine(T)
                            End If
                        Next
                    Next
                Next
            End Sub

        End Class

    End Class

End Namespace

File: Common/Utilities.Reflection.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Runtime.Loader;

using Reflect = System.Reflection;

namespace Ks.Common
{
    partial class Utilities
    {
        public static class Reflection
        {

            public static IEnumerable<Reflect.Assembly> GetAllAccessibleAssemblies()
            {
                return Enumerable.Concat(Reflect.Assembly.GetEntryAssembly().GetRecursiveReferencedAssemblies(), AssemblyLoadContext.All.SelectMany(alc => alc.Assemblies)).Distinct();
            }

            public static IEnumerable<Reflect.MethodInfo> GetAllMethods(IEnumerable<Reflect.Assembly> Assemblies = null)
            {
                if (Assemblies == null)
                {
                    Assemblies = GetAllAccessibleAssemblies();
                }

                foreach (var Ass in Assemblies)
                {
                    foreach (var Type in Ass.GetTypes())
                    {
                        foreach (var Method in Type.GetMethods(Reflect.BindingFlags.Static | Reflect.BindingFlags.Instance | Reflect.BindingFlags.Public | Reflect.BindingFlags.NonPublic))
                        {
                            yield return Method;
                        }
                    }
                }
            }

            public static IEnumerable<Type> GetAllTypes(IEnumerable<Reflect.Assembly> Assemblies = null)
            {
                if (Assemblies == null)
                {
                    Assemblies = GetAllAccessibleAssemblies();
                }

                foreach (var Ass in Assemblies)
                {
                    foreach (var Type in Ass.GetTypes())
                    {
                        yield return Type;
                    }
                }
            }

            public static IEnumerable<Type> GetAllTypesDerivedFrom(Type Base, IEnumerable<Reflect.Assembly> Assemblies = null)
            {
                if (Assemblies == null)
                {
                    Assemblies = GetAllAccessibleAssemblies();
                }

                Base = Base.GetGenericTypeDef();

                foreach (var Type in GetAllTypes(Assemblies))
                {
                    if (Type.GetBaseTypes().Concat(Type.GetInterfaces()).Select(t => t.GetGenericTypeDef()).Contains(Base))
                    {
                        yield return Type;
                    }
                }
            }
        }
    }
}
